### PR TITLE
Improve Array Memory Management

### PR DESCRIPTION
**Pull request by mppf**
## Overview

This PR re-implements memory management for arrays, domains, and distributions. The changes mainly fall into these groups:
 * array-related compiler changes:
 * tuple changes
 * other enabling changes in the compiler
 * changes to module code for arrays/domains/distributions
 * other module code changes
 * test changes

## User-facing behavior changes

This PR causes mainly four behavior changes:

 * Arrays are returned by value
 * Arrays, domains, and distributions are no longer kept alive (preventing deletion) by a task using them
 * Arrays are no longer kept alive by slices
 * Tuple behavior is refined, especially for blank-intent tuple arguments

The first of these is a new decision for the language in order to make it more implementable and to make arrays more consistent with records and other types. The second and third are implied by the Lexical Scoping section in the Chapel Evolution document (http://chapel.cray.com/docs/latest/language/evolution.html#lexical-scoping). The last addresses a known design issue in the language and is discussed more below.

Note that while arrays return by value by default, the ref-return-intent or const-ref-return-intent can override that behavior.

In order to make these changes more achievable, it is not currently defined what happens when returning a slice from a function. Currently, such slices return without making a copy - and so it is possible to create core dumps by returning a slice of a local array.

The tuple behavior changes adjust the implementation so that a 1-element tuple behaves similarly to just the element (as far as whether or not a copy occurs). Then the behavior logically extends to tuples with more elements. In particular, creating a tuple containing records and then passing it to a function does not copy the record. But returning such a tuple formal argument or using it to initialize a variable will copy the record. This design has similar goals and ideas to the design presented in CHIP 6 but differs slightly in the details.

## Array-Related Compiler Changes

- removing cases when an array wrapper record is shallow-copied in the compiler. Either the copy constructor should be called (which does a deep copy) or the variable in question should be a reference to the array (which does no copy and does not impact array lifetime). Now autoCopy and initCopy do the same thing for arrays, domains, and distributions. E.g. returning an _array causes the array to be returned by value, but returning a _ref(_array) causes it to be returned by reference.
- adjusts the compiler to treat arrays, domains, and distributions similarly to user records and strings.
- adds a workaround to make `var x => GlobalArray[1..10]` work
- disables most of removeUnnecessaryAutoCopyCalls. Only autoCopy calls for POD types are removed. The rest of this pass should no longer be necessary since we now just do a better job at inserting the minimal number of autoCopies in the first place.
- disables removeWrapRecords
- remove _checkIterator because it is unnecessary and was causing problems with arrays
- normalize: remove special treatment for iterators with declared array return type
- parallel: remove special handling for arrays
- addAutoDestroyCalls: check for FLAG_NO_AUTO_DESTROY before adding an autoDestroy call. This flag is useful in some special code to do with arrays/domains/distributions.
- cullOverReferences: delete code that would remove references to array/domain/dist wrapper records since now the difference between _array and _ref(_array) is critical.
- (see description of array module changes) changed _array/_domain/_distribution field _valueType to _instance.
- wrappers: adjustments to array alias fields in constructors. Also set argument type for promotion wrappers to array reference types.
- in callDestructors, adds fixupNewAlias to fix up certain cases using the array alias operator, such as returning an alias of a slice expression.

## Tuple Changes

- re-implementing tuple functions in the compiler so that they can be adjusted appropriately. The problem was that _build_tuple was instantiated by varargs processing and was too hard to adjust in that context.
- now the tuple constructor calls initCopy for elements captured by value.
- now tuples capture arrays and records by reference in some situations and by value in others. The guiding principle here is that a 1-element tuple should behave similarly to just the element. So a tuple passed by blank intent captures arrays and records by ref, since these are passed by reference when passed by blank intent. A returned tuple captures arrays and records by value, since these are returned by value.
- tuple functions to be filled in by the compiler have flags in some cases to identify them and are saved into global variables by filesToAST or in scopeResolve if the relevant functions are created there.
- tuples now have an 'unref' function which is used to implement converting a tuple from a tuple containing references to one containing only values when returning.
- in getNewSubType, rewrote it to be clearer what is going on

## Other Enabling Changes in the Compiler

enabling ref changes:

  - deadCodeElimination: isDeadVariable checks uses ->isRef instead of checking flag
  - removeValueForwarding: forward array/domain/distribution values
  - insertWideReferences: fixes for references
  - parallel: adjust makeHeapAllocations not to remove autoDestroy functions and to not make copies of arguments passed by blank-intent-is-ref.

iterator changes:

  - removing _checkIterator
  - iterators capture arrays, domains, distributions by reference (when
    creating the iterator record)

## Changes to module code for arrays/domains/distributions

- _array/_domain/_distribution now have _instance and _pid fields instead of _value and _valueType fields. This enables destruction of privatized objects since we can tell them apart (see _freePrivatizedClass). Previously, the _value field was sometimes an integer pid and sometimes a pointer to the instance.
- adds dsiDestroyDist/dsiDestroyDom/dsiDestroyArr
- moves pid field from DSI implementors to BaseDist/BaseDom/BaseArr
- adds dsiSingleton, dsiMyDist
- removes _value and _valueType from _distribution/_domain/_array; adds _pid, _instance, and _unowned fields instead
- removes BaseDist/Dom/Arr _distCnt/_domCnt/_arrCnt. There is no array reference counter now. Domain and distribution generally use the number of elements in their list of arrays and domains respectively. Domains also have a counter _arrs_containing_dom, but this only is used for a bug fix for arrays-of-arrays.
- Since destroying an array/domain/dist can cause other objects to need to be destroyed, but the implementation (with the DSI interface) might need to take action on the destruction, the process of destruction generally works in two phases: 1) remove objects from the tree (e.g. removing a domain from its distribution) and figure out any follow-on objects that must be destroyed (e.g. when removing the last domain in a distribution). 2) call dsiDestroy on the destroyed object. 3) destroy any follow-on objects and call dsiDestroy on them too in the process.
- As a work-around for apparent compiler bugs, added empty destructors for many array/domain/distribution implementation classes. A TODO is to try removing these empty destructors and fix the compiler bug.

## Other Changes to Module code

- changes to track tuple re-implementation
- removal of _checkIterator
- simplified locale initialization by using a DummyLocale class for 'here' while the locale hierarchy is being initialized.

## Test Changes

-  removing the "assignment has been serialized" warning in many tests. The reason is that the warning is no longer useful because it occurs any time e.g. associative array variables are declared. That is now happening because the autoCopy function for these arrays is being resolved since the compiler might add calls to it, and now this autoCopy function can perform a deep copy.
- add quite a few array semantics & leak checks in test/arrays/ferguson/semantic-examples
- new tests of new tuple behavior in test/types/tuple/ferguson
- other minor changes to tests; see individual commit messages

## Testing

Passes quickstart testing.
Passed quickstart --no-local testing.
Passed full local testing.
Passed GASNet + qthreads testing.

## Review

Reviewed by @lydia-duncan and @vasslitvinov - thanks!

## TO-DOs

Minor implementation improvements:

* remove the param noRefCount & update performance tests that throw it
  (after this PR, noRefCount exists but has no effect)
* consistetly use nullPid instead of -1
* explore storing the pid field only in the wrapper record or only in the
  pointed to class (e.g. BaseArr)
* explore removing domain.linksDistribution in favor of distribution.trackDomains.
* find and fix the compiler error worked-around by the empty destructors
* consider only creating chpl__unalias for arrays (instead of it doing nothing for other types)
* clean-up and/or remove removeUnnecessaryAutoCopyCalls.cpp
* investigate why specifying the return type in ReplicatedArr.dsiAccess caused a problem
* investigate comm counts regressions and multilocale performance.

Bigger implementation improvements:
* fix the memory leak when initializing generic fields (which causes
  ReplicatedDist to leak)
* remove the distinction between initCopy and autoCopy
* make initCopy / autoCopy call record copy-initializer
* rename initCopy / autoCopy to something else
* Optimize in, out, and inout argument intents to remove unnecessary copying
* Optimize record constructors to remove unnecessary copying

Documentation improvements:

* describe function return (by value) in the language specification
* describe when copies occur in the language specification
* describe tuple semantics in the language specification

Design improvements:

* decide on the default argument intent for ranges and how to pass ranges into begin functions (see workaround in functionResolution.cpp).
* finalize and document tuple ref-ness rules
* decide whether or not it is an error to return a slice of a local array. If appropriate, produce an error message during compilation for programs that return a slice of a local array
* Decide whether or not chpl__unref and chpl__unalias hooks are available to users

### CODE REVIEW

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -6335,26 +6645,70 @@ preFold(Expr* expr) {
 
       result = paramLoop->foldForResolve();
     } else if (call->isPrimitive(PRIM_LOGICAL_FOLDER)) {
-      bool removed = false;
+
+      if(call->parentSymbol &&
+          call->parentSymbol->id == 950612)
+        gdbShouldBreakHere();
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: You'll want to remove this, I think it's leftover debugging stuff?

Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: yes, good catch

File: compiler/main/checks.cpp [hunk 2]
```diff
@@ -470,6 +470,15 @@ static void check_afterResolveIntents()
       // Only look at Var or Arg symbols
       if (isLcnSymbol(sym)) {
         QualifiedType qual = sym->qualType();
+        // MPF TODO: This should not be necessary
+        // it is a workaround for problems with --verify
+        // with tuple type constructors accepting domains.
+        // It would be better to treat run-time types a
+        // normal record.
```
Comment (lydia-duncan) [inline compiler/main/checks.cpp]: Typo in this comment.  Should be "better to treat runtime types *as* normal *records*"?

File: compiler/resolution/functionResolution.cpp [hunk 3]
```diff
@@ -4565,10 +4812,37 @@ static void resolveMove(CallExpr* call) {
...
+  // QualifiedType.
+  if (lhs->hasFlag(FLAG_MAYBE_REF) && !isReferenceType(rhsType)) {
+    if (SymExpr* se = toSymExpr(rhs)) {
+      if (ArgSymbol* arg = toArgSymbol(se->var)) {
+        if (concreteIntent(arg->intent, arg->type) & INTENT_FLAG_REF) {
+          makeRefType(rhsType);
+          rhsType = rhsType->refType;
+
+          // Add PRIM_ADDR_OF
+          //  (this won't be necessary with QualifiiedType/PRIM_SET_REFERENCE)
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: Typo on "QualifiiedType"

File: test/arrays/ferguson/if-expr/aif-param.chpl [hunk 4]
```diff
@@ -0,0 +1,21 @@
+config param param_useA = true;
```
Comment (lydia-duncan) [inline test/arrays/ferguson/if-expr/aif-param.chpl]: I'm having trouble parsing the name of this file (and the similarly named ones).  Maybe add a comment at the top of what it means?

Comment (lydia-duncan) [inline test/arrays/ferguson/if-expr/aif-param.chpl]: I think these are good tests, though

File: compiler/optimizations/copyPropagation.cpp [hunk 5]
```diff
@@ -213,6 +213,16 @@ static void extractReferences(Expr* expr,
                      lhs->name, lhs->id, val->name, val->id);
 #endif
             refs.insert(RefMapElem(lhs, val));
+          } else {
+#if DEBUG_CP
+            if (debug > 0) {
+              printf("Setting pair to NULL: %s[%d]\n", lhs->name, lhs->id);
+            }
+#endif
+            // If we can't reason about this useage of a reference, mark it's
```
Comment (lydia-duncan) [inline compiler/optimizations/copyPropagation.cpp]: Typo on "useage" and I believe that's the wrong "it's"

**Comment by lydia-duncan** [comment]
I am now up to date on your changes again

**Comment by mppf** [comment]
Known performance regressions:
 * op= with --no-local
 * miniMD and STREAM on 16-node XC

**Pull request merged**